transports/quic/src/muxer.rs

```diff
@@ -0,0 +1,360 @@
+        let substream = Substream::new(substream_id, self.inner.clone());
+        Poll::Ready(Ok(substream))
+    }
+
+    fn poll_outbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.inner.lock();
+        let substream_id = match inner.connection.open_substream() {
+            Some(id) => {
+                inner.poll_outbound_waker = None;
+                id
+            }
+            None => {
+                inner.poll_outbound_waker = Some(cx.waker().clone());
+                return Poll::Pending;
+            }
+        };
+        inner.substreams.insert(substream_id, Default::default());
+        let substream = Substream::new(substream_id, self.inner.clone());
+        Poll::Ready(Ok(substream))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let Inner {
+            substreams,
+            connection,
+            ..
+        } = &mut *self.inner.lock();
+        if connection.is_drained() {
+            return Poll::Ready(Ok(()));
+        }
+
+        if connection.send_stream_count() != 0 {
+            for substream in substreams.keys() {
+                if let Err(e) = connection.finish_substream(*substream) {
+                    log::warn!("substream finish error on muxer close: {}", e);
+                }
+            }
+        }
+        loop {
+            if connection.send_stream_count() == 0 && !connection.is_closed() {
+                connection.close()
+            }
+            match connection.poll_event(cx) {
+                Poll::Ready(ConnectionEvent::ConnectionLost(_)) => return Poll::Ready(Ok(())),
+                Poll::Ready(_) => {}
+                Poll::Pending => break,
+            }
+        }
+        Poll::Pending
+    }
+}
+
+pub struct Substream {
+    id: quinn_proto::StreamId,
+    muxer: Arc<Mutex<Inner>>,
+}
+
+impl Substream {
+    fn new(id: quinn_proto::StreamId, muxer: Arc<Mutex<Inner>>) -> Self {
+        Self { id, muxer }
+    }
+}
+
+impl AsyncRead for Substream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        mut buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let mut muxer = self.muxer.lock();
+
+        let mut stream = muxer.connection.recv_stream(self.id);
+        let mut chunks = match stream.read(true) {
+            Ok(chunks) => chunks,
+            Err(quinn_proto::ReadableError::UnknownStream) => {
+                return Poll::Ready(Ok(0));
+            }
+            Err(quinn_proto::ReadableError::IllegalOrderedRead) => {
+                unreachable!(
+                    "Illegal ordered read can only happen if `stream.read(false)` is used."
+                );
+            }
+        };
+        let mut bytes = 0;
+        let mut pending = false;
+        loop {
+            if buf.is_empty() {
+                break;
+            }
```

Comment:
This condition being necessary seems to be extremely subtle. From reading the code, I gather that the issue is that `Chunks::next` will happily take a `max_length` or `0` in which case we will loop forever here, reading 0 sized chunks :upside_down_face:.

We should either add a comment for this or add `if !chunk.bytes.is_empty()` as a match-arm guard to the `Ok(Some)` branch.

Reply:
We only request chunks of length `buf.len()` from the quinn API (line 261), so ideally, that would just return `Ok(None)` if we request a 0-sized chunk. But it doesn't. It returns a `Chunk` of length `0` which means we will loop forever requesting new chunks and never break.

Reply:
Ah I see. But checking if we have space to even write into the buf seems reasonable to me. And you don't need to know anything about how things are chunked.

Reply:
Included match-arm guard in 82b6a9d.